### PR TITLE
Avatar-UI : giving a white background and cornering the status Indicator

### PR DESCRIPTION
**Pull request by abhi0504**
## **User's Avatar**

I have made 2 changes majorly

1. Gives a white background larger than the status indicator due to which no matter what will be the Avatar Image of the user the status indicator looks clear .
for example If the user's Avatar Image is greenish then the present indicator not looks that well 
Also above point can be easily understand by Greg's Issue #4306 

2. We need to check whether the status is offline or Its a Zulip's Bot and if it is then there is no need to add any white background as no status indicator is there but If we doesn't check then we get a white background over there 
                                                                              
### _not considering that If case_
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_58_02 (2)          Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_57_38 (2)

### _considering that If case_
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_58_23 (2)          Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_58_33 (2)

So after the above 2 changes Avatars of user's in User's List looks like this

### Before 
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_35 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_34_23 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_33_59 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_55 (2)

### After
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_22_58 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_23_27 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_25_15 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_26_06 (2)

Fixes : #4306

### CODE REVIEW

**Comment by abhi0504** [comment]
@chrisbobbe plz review this PR

**Comment by gnprice** [comment]
Hi @abhi0504! Here's some comments from a quick look:
* Let's avoid duplicating here any of the logic from `PresenceStatusIndicator` that figures out what kind of status indicator is going to show.
  * Instead, let's have `PresenceStatusIndicator` itself provide this background. This would be controlled by a boolean prop so that it's enabled here but not in other places we use `PresenceStatusIndicator` where we might not want the background.
* The specific mention of `email === 'welcome-bot@zulip.com'` is a bad sign -- that is definitely not going to be the right logic for this. Whatever's going on for welcome-bot is going to be true of some other users too, so let's take the time to figure that out so we can have the right logic to cover them. (Similarly, it might be true of welcome-bot in your testing now but not true of welcome-bot in some other circumstances.)
* In your comment here on the issue: https://github.com/zulip/zulip-mobile/issues/4306#issuecomment-727253127 you had some screenshots where the status icons looked somewhat different from these. Is that from a previous version of your changes, or are there some places where it looks like that and others where it looks like your screenshots above?
* In the screenshots above, the white borders look pretty thick. Let's try them thinner -- like about the same thickness as the gray or orange borders in two of the status icons.

**Comment by abhi0504** [comment]
Hi @gnprice Thanks for your comment , 
I am making a new commit which will clear all above points , this commit will do following things :

- Now Instead of duplicating the logic from `PresenceStatusIndicator` in `UserAvatarWithPresence` I am using a boolean prop `isUserAvatar` from `UserAvatarWithPresence` which tells whether `PresenceStatusIndicator` is used for User Avatar or not and if Its true then we are returning with that extra `View` which gives it a white background.

- Now there is also no need of specific mention of `email === 'welcome-bot@zulip.com'` as we are returning that white background only in the case of `IDLE` and `ACTIVE` user 
- In my comment here on the issue:[#4306](https://github.com/zulip/zulip-mobile/issues/4306#issuecomment-727253127) the screenshots that I have mentioned are from my previous version of changes , and the screen shots present above (in the current PR) are the main screen shots 
- Yes In the above screen shot the border is more thicker than the previous one which in my opinion looks good and gives it notch like feel but if you would save I can decrease its thickness like the [previous ones](https://github.com/zulip/zulip-mobile/issues/4306#issuecomment-727253127) or like about the same thickness as the gray or orange borders in two of the status icons as you have mentioned plz reply what looks good here 
- Also If you would say then I will squash all the commits into One by which this PR have only one final commit in it before merging

File: src/common/UserAvatarWithPresence.js [hunk 1]
```diff
@@ -12,8 +12,8 @@ import PresenceStatusIndicator from './PresenceStatusIndicator';
 
 const styles = createStyleSheet({
   status: {
-    bottom: 0,
-    right: 0,
+    bottom: 4,
+    right: 4,
```
Comment (chrisbobbe) [inline src/common/UserAvatarWithPresence.js:16]:
```diff
-    bottom: 0,
-    right: 0,
+    bottom: 4,
+    right: 4,
```

Hmm, what's going on here? I haven't yet run the app with this revision, but it's not totally clear how this small code change contributes to the behavior change described in the commit message.

File: src/common/PresenceStatusIndicator.js [hunk 2]
```diff
@@ -110,6 +128,22 @@ class PresenceStatusIndicator extends PureComponent<Props> {
       return null;
     }
 
+    if (isUserAvatar && status === 'active') {
```
Comment (chrisbobbe) [inline src/common/PresenceStatusIndicator.js:131]: Hmm, what's the reason that `isUserAvatar` is only checked for the `active` and `idle` statuses? How about `offline` and `unavailable` too?

**Review by chrisbobbe — CHANGES_REQUESTED** [review]
Hi @abhi0504! Here are some comments from a quick look at the code just now.

File: src/common/PresenceStatusIndicator.js [hunk 3]
```diff
@@ -17,6 +17,15 @@ const styles = createStyleSheet({
     height: 12,
     borderRadius: 6,
   },
+  commonbkg: {
```
[the remark this answers is not in the document]
Comment (abhi0504) [reply src/common/PresenceStatusIndicator.js:20]: sure I will changed it from commonbkg to opaqueBackground

File: src/common/PresenceStatusIndicator.js [hunk 4]
```diff
@@ -81,6 +90,7 @@ type Props = $ReadOnly<{|
   style?: ViewStyleProp,
   email: string,
   hideIfOffline: boolean,
+  isUserAvatar: boolean,
```
[the remark this answers is not in the document]
Comment (abhi0504) [reply src/common/PresenceStatusIndicator.js:93]: sure I will changed it from isUserAvatar to useOpaqueBackground

File: src/common/UserAvatarWithPresence.js (hunk 1 above)
Comment (abhi0504) [reply src/common/UserAvatarWithPresence.js:13]: this changes helps us to centralize the status indicator on the white background

File: src/common/PresenceStatusIndicator.js (hunk 2 above)
Comment (abhi0504) [reply src/common/PresenceStatusIndicator.js:131]: for offline and unavailable statuses there is no need of white background required

**Comment by abhi0504** [comment]
@gnprice and @chrisbobbe should I squash that second commit into first one or its ok for now. Also plz review the above changes

File: src/common/PresenceStatusIndicator.js (hunk 2 above)
Comment (chrisbobbe) [reply src/common/PresenceStatusIndicator.js]: Let's do a white background for all four statuses. Like you've done in your "after" screenshots from https://github.com/zulip/zulip-mobile/pull/4324#issue-530742703, right? (Or, if not: can you say why you don't want to do it for "offline" or "unavailable" anymore? 🙂) Quoting from your message:

> ### Before
> Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_35 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_34_23 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_33_59 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_55 (2)
> 
> ### After
> Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_22_58 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_23_27 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_25_15 (2) Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_26_06 (2)

Under "After", I see an opaque white background for all status indicators.

-----

I noticed that, in this revision, the "unavailable" indicator still doesn't stand out very well:

image

Also, the "offline" indicator, when it appears—i.e., when `PresenceStatusIndicator` gets `false` for its `hideIfOffline` prop—doesn't stand out quite enough when the colors are similar. (I do see that `UserAvatarWithPresence` gives `true` for `hideIfOffline`. But `PresenceStatusIndicator` should do the right thing for all inputs it claims to accept.)

image

**Comment by chrisbobbe** [comment]
> @gnprice and @chrisbobbe should I squash that second commit into first one or its ok for now.

Yes, please do squash the second commit into the first one. 😉

**Review by chrisbobbe — CHANGES_REQUESTED** [review]
Thanks, @abhi0504! See some comments below, and another one [above](https://github.com/zulip/zulip-mobile/pull/4324#discussion_r551460168).

I'd also like to see a version with thinner borders; I like Greg's idea from https://github.com/zulip/zulip-mobile/pull/4324#issuecomment-745128290:
> about the same thickness as the gray or orange borders in two of the status icons

**Comment by abhi0504** [comment]
@chrisbobbe Thank You so much for review .
So the things is we have two different versions of borders thinner and broader screenshot of both the versions are below
### Before The PR
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_35 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_34_23 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_33_59 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_32_55 (2)

### Broader or thick Borders
Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_22_58 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_23_27 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_25_15 (2)    Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_26_06 (2)

### Thinner Borders
Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_07_59     Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_08_51     Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_10_09     Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_10_36

The thing is If we have thin borders (the same thickness as the gray or orange borders) then there is no need of white Background for the   "Active" and "Offline" status as they are fully covered by status indicator .

So as screenshots for both the cases are above plz tell what looks good I will commit accordingly and squash all commits into one single commit .

**Comment by chrisbobbe** [comment]
> The thing is If we have thin borders (the same thickness as the gray or orange borders) then there is no need of white Background for the "Active" and "Offline" status as they are fully covered by status indicator .

I think you might be confusing "thin borders" with "no borders"; I don't see any white border in any of the four screenshots under "Thinner borders".

I'm thinking thinner than this border

Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_26_06 (2)

But not so thin that the border disappears (as it does here):

Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_10_09

You could try a width of 1.5 or 2; I'm getting those numbers from `idleWrapper` and `unavailableWrapper`, which look like they control the width of the orange and gray borders in two of the status indicators.

**Comment by abhi0504** [comment]
> > The thing is If we have thin borders (the same thickness as the gray or orange borders) then there is no need of white Background for the "Active" and "Offline" status as they are fully covered by status indicator .
> 
> I think you might be confusing "thin borders" with "no borders"; I don't see any white border in any of the four screenshots under "Thinner borders".
> 
> I'm thinking thinner than this border
> 
> Android Emulator - Pixel_3_API_29_5554 02-12-2020 09_26_06 (2)
> 
> But not so thin that the border disappears (as it does here):
> 
> Android Emulator - Pixel_3_API_29_5554 05-01-2021 01_10_09
> 
> You could try a width of 1.5 or 2; I'm getting those numbers from `styles.idleWrapper` and `styles.unavailableWrapper`, which look like they control the width of the orange and gray borders in two of the status indicators.

Oh Okay Now I understood Thanks @chrisbobbe 
I am making borders a little more broader than the last one also adding the screenshots of all four cases below 
Android Emulator - Pixel_3_API_29_5554 05-01-2021 16_18_31 (2)     Android Emulator - Pixel_3_API_29_5554 05-01-2021 20_48_14 (2)     Android Emulator - Pixel_3_API_29_5554 05-01-2021 20_59_33 (2)     Android Emulator - Pixel_3_API_29_5554 05-01-2021 21_00_30 (2)

Also You have mentioned some changes above which are also done and I squash all commits into one kindly plz review The changes and PR .

**Comment by chrisbobbe** [comment]
Thanks, @abhi0504! I like the new thickness of the borders.

Merged, with a revised commit message and a few [NFC](https://github.com/zulip/zulip-mobile/blob/570db381dd518c37dba61f715ee35b4e3a89c35b/docs/glossary.md#nfc-or-nfc) commits on top of yours; please take a look! 🙂 The code at your commit works, but I saw that we could tighten up the logic by making it easier to see what precisely in `PresenceStatusIndicator` was controlled by `useOpaqueBackground`.

**Pull request merged**